### PR TITLE
Start to move away from complex and imaginary types

### PR DESCRIPTION
**Pull request by wilzbach**
This is first PR which works in the direction of finally going through with the
complex and imaginary type deprecation.

See also: https://github.com/dlang/dmd/pull/7640

### CODE REVIEW

File: std/format.d [hunk 1]
```diff
@@ -5425,37 +5425,23 @@ enum Mangle : char
...
-    }
-    auto p = m in dic;
-    return p ? *p : null;
+    enum types = [
+        "void", "bool", "byte", "ubyte",
+        "short", "ushort", "int", "uint", "long", "ulong",
+        "float", "double", "real",
+        "char", "wchar", "dchar",
+        "ifloat", "idouble", "ireal",
+        "cfloat", "cdouble", "creal",
```
Comment (wilzbach) [inline std/format.d]: How can we call `typeid` without a deprecation warning?

File: std/math.d [hunk 2]
```diff
@@ -526,50 +555,48 @@ enum real SQRT1_2 =    SQRT2/2;                               /** $(SQRT)$(HALF)
...
-}
-
 /// ditto
 @safe pure nothrow @nogc unittest
 {
+    import std.complex : complex;
     assert(isIdentical(abs(-0.0L), 0.0L));
     assert(isNaN(abs(real.nan)));
     assert(abs(-real.infinity) == real.infinity);
-    assert(abs(-3.2Li) == 3.2L);
```
Comment (wilzbach) [inline std/math.d]: The complex deprecation currently triggers error in deprecated code. See https://issues.dlang.org/show_bug.cgi?id=18212

File: std/format.d (hunk 1 above)
Comment (ibuclaw) [reply std/format.d]: I think the right steps are to first deprecate in library (user-code).  Once enough time has passed, remove support for complex and imaginary in the library.

Then we kill it in the compiler.

**Comment by wilzbach** [comment]
> I think the right steps are to first deprecate in library (user-code). Once enough time has passed, remove support for complex and imaginary in the library.
> Then we kill it in the compiler starting with deprecation by default.

We can't do that. Phobos needs to build without deprecation messages. This was enabled about 9 months ago, because (1) deprecation messages were shown to users and (2) people always introduced deprecations without considering the fallout and viability. By `-de` we can enforce that if a symbol gets deprecated in Phobos, all usages of it have been removed (or are in `deprecated` blocks). Since it was introduced, it has proven as quite helfpul.

**Comment by andralex** [comment]
>Phobos needs to build without deprecation messages.

Slight amendment: it's okay to have deprecation messages in phobos' unittest builds.

**Comment by wilzbach** [comment]
> Slight amendment: it's okay to have deprecation messages in phobos' unittest builds.

Yep, but only if they are marked as `deprecated`. However, the complex deprecation messages doesn't respect this at the moment: https://issues.dlang.org/show_bug.cgi?id=18212

File: std/math.d [hunk 3]
```diff
@@ -2545,42 +2667,45 @@ private real exp2Impl(real x) @nogc @trusted pure nothrow
...
-                fsincos;
-                fxch ST(1), ST(0);
-            }
-        }
-    }
-    else
-    {
-        return cos(y) + sin(y)*1i;
-    }
+auto expi(real y) @trusted pure nothrow @nogc
```
Comment (wilzbach) [inline std/math.d]: Not sure how to deal with this in another way.

File: std/math.d [hunk 4]
```diff
@@ -732,50 +818,72 @@ float sin(float x) @safe pure nothrow @nogc { return sin(cast(real) x); }
...
+    assert(sin(complex(2.0, 0.0)) == sin(2.0L));
 }
 
 /***********************************
  *  cosine, complex and imaginary
  *
  *  cos(z) = cos(z.re)*cosh(z.im) - sin(z.re)*sinh(z.im)i
  */
-creal cos(creal z) @safe pure nothrow @nogc
+auto cos(C)(C z) @safe pure nothrow @nogc
```
Comment (wilzbach) [inline std/math.d]: Not sure how to deal with this in another way.

File: std/format.d [hunk 5]
```diff
@@ -5419,31 +5420,6 @@ enum Mangle : char
     Timmutable = 'y',
 }
 
-// return the TypeInfo for a primitive type and null otherwise.  This
-// is required since for arrays of ints we only have the mangled char
-// to work from. If arrays always subclassed TypeInfo_Array this
-// routine could go away.
-private TypeInfo primitiveTypeInfo(Mangle m)
```
Comment (wilzbach) [inline std/format.d]: So `primitiveTypeInfo` is a relict from the old `doFormat` days. It's not used anywhere anymore.
See e.g. https://github.com/dlang/phobos/pull/4978

File: std/math.d [hunk 6]
```diff
@@ -507,30 +504,38 @@ enum real SQRT2 =      0x1.6a09e667f3bcc908b2fb1366ea958p+0L; /** $(SQRT)2 = 1.4
 enum real SQRT1_2 =    SQRT2/2;                               /** $(SQRT)$(HALF) = 0.707106... */
 // Note: Make sure the magic numbers in compiler backend for x87 match these.
 
+// it's quite tricky check for a type who will trigger a deprecation when accessed
+template isDeprecatedComplex(T)
+{
+    static if (__traits(isDeprecated, T))
```
Comment (wilzbach) [inline std/math.d]: Requires https://github.com/dlang/dmd/pull/7665

File: std/complex.d [hunk 7]
```diff
@@ -874,16 +876,61 @@ Complex!real expi(real y)  @trusted pure nothrow @nogc
...
 
+/**
+    Params: y = A real number.
+    Returns: The value of cosh(y) + i sinh(y)
+
+    Note:
+    $(D coshisinh) is included here for convenience and for easy migration of code
+    that uses $(REF _coshisinh, std,math).
+*/
+Complex!real coshisinh(real y) @safe pure nothrow @nogc
```
Comment (wilzbach) [inline std/complex.d]: This is part of https://github.com/dlang/phobos/pull/5698

**Comment by wilzbach** [comment]
So it was quite to tricky to get Phobos to build, but thanks to https://github.com/dlang/dmd/pull/7647
we can now "poison" deprecate functions and the diff got a lot cleaner.

-------

BTW it was quite tricky to get `abs` working as every usage of the complex type will rightfully trigger a deprecation warning, even hacks like this:

```d
pragma(msg, creal.mangleof);
```
https://run.dlang.io/is/iNPtWM

However, there's the new `isDeprecated` traits which is exactly intended for such purposes. It just needs to be made aware of the complex types: https://github.com/dlang/dmd/pull/7665

**Comment by wilzbach** [comment]
So while this is slowly falling into the abyss of the universe, I completed the quest of finding the unittests using complex or imaginary numbers. It turns out, there aren't too many, but as some are in `std.format` it still lead to very long deprecation messages.

Anyhow we can now use `-transition=complex` on Phobos for everything (build, unittest) and once this is merged should be able to finally turn the switch at DMD :)

However, while doing this I hit two more bugs:
- https://issues.dlang.org/show_bug.cgi?id=18250
- https://issues.dlang.org/show_bug.cgi?id=18249

The problem is that deprecated works poorly with templates, so it doesn't mark a template as deprecated when instantiated:

```d
auto foo(T)(T t) // <- deprecation message
{
    return T.init;
}

deprecated unittest
{
    auto b = foo(2 + 2i);
}
``` 
// 18250

and more annoyingly even does this if constraint evaluates to false.

```
auto foo(T)(T t) // <- deprecation message
if (!__traits(isDeprecated, T))
{
    return T.init;
}

deprecated
auto foo(T)(T t)
if (__traits(isDeprecated, T))
{
    return T.init;
}

deprecated unittest
{
    auto b = foo(2 + 2i);
}
```

The underlying problem is that the `DSymbolSemanticVisitor` just checks its in the `VarDeclaration` overload whether the type of the currently visited `Var` is complex/imaginary:

https://github.com/dlang/dmd/blob/11eda055563457560ff2510048805880ed8d8c28/src/dmd/dsymbolsem.d#L385

File: std/format.d [hunk 8]
```diff
@@ -2604,6 +2605,8 @@ if (is(Unqual!T : creal) && !is(T == enum) && !hasToString!(T, Char))
     put(w, 'i');
 }
 
+version(TestComplex)
```
Comment (wilzbach) [inline std/format.d]: https://issues.dlang.org/show_bug.cgi?id=18250 - I didn't know a better way to workaround it.

File: std/format.d [hunk 9]
```diff
@@ -2621,6 +2624,8 @@ if (is(Unqual!T : creal) && !is(T == enum) && !hasToString!(T, Char))
     }
 }
 
+version(TestComplex)
```
Comment (wilzbach) [inline std/format.d]: https://issues.dlang.org/show_bug.cgi?id=18250 - I didn't know a better way to workaround it.

File: std/math.d [hunk 10]
```diff
@@ -543,20 +563,22 @@ if ((is(typeof(Num.init >= 0)) && is(typeof(-Num.init)) ||
     }
 }
 
-/// ditto
-auto abs(Num)(Num z) @safe pure nothrow @nogc
-if (is(Num* : const(cfloat*)) || is(Num* : const(cdouble*))
-    || is(Num* : const(creal*)))
+import std.meta : AliasSeq;
+deprecated("Please use std.complex")
+static foreach (Num; AliasSeq!(cfloat, cdouble, creal, ifloat, idouble, ireal))
```
Comment (wilzbach) [inline std/math.d]: This is pretty ugly, but my attempt at working around https://issues.dlang.org/show_bug.cgi?id=18251

**Comment by wilzbach** [comment]
Meh. https://github.com/dlang/druntime/pull/2047 is needed first :/

**Comment by 9il** [comment]
Please do not deprecate native complex numbers!

std.complex does not work with LDC optimisation flags. I would need to copy std.complex to mir, mark with appropriate LDC flags, add my own conditional compilation flags, and add this informations to few readmes.

Native complex numbers are easy to use, they do not require Phobos and they works as _native_ . This means that `a + b` is not a function. And this is very important for LDC optimization flags. I do not want to copy phobos module again :-(

Complex numbers can be implemented as library but LDC should inline this function in front-end, not it LLVM back-end. Otherwise they would be much slooooower. Proved in GLAS&

**Comment by andralex** [comment]
@9il I'm interested in learning more about this. Can't compilers inline `a + b` as well as a builtin would do?

**Comment by ibuclaw** [comment]
@9il - slow down. Say one thing at a time, and be concise. Your arguments either make no sense or are conflicting.

**Comment by thewilsonator** [comment]
@ibuclaw I get what @9il is saying

He means that `std.complex` is not annotated with LLVM optimisation hints (inlining, fast math). This leads to either duplication, or performance penalties and a dependance on phobos. Neither of which is desirable.

@andralex it should be possible to hook up the phobos symbols to the current code generating machinery, but this doesn't solve the dependance on phobos.

The simplest option would be to annotate phobos (with `version(LDC)` or in LDC's copy of phobos) with optimisation hints and since everything except `expi` is a template that shouldn't introduce a dependance on phobos if it is made to be a template.

The only remaining issue is the use of `real` (e.g. in `std.complex.sqrt`), which is very undesirable when speed is of greatest concern.

**Comment by andralex** [comment]
First order of business (assuming we can implement complex as a lib) seems to move fundamental complex support into druntime. Code using complex numbers should only depend on core support.

>The only remaining issue is the use of real (e.g. in std.complex.sqrt), which is very undesirable when speed is of greatest concern.

This should be addressed separately, anyone working on this?

**Comment by thewilsonator** [comment]
>First order of business (assuming we can implement complex as a lib) seems to move fundamental complex support into druntime. Code using complex numbers should only depend on core support.

It will need to work with `-betterC` (i.e. not depend on `libdruntime.a`)to have any hope of making Ilya happy. But assuming that all of the basic operations are `pragma(inline, true)`, I don't think this will be a problem. 
> This should be addressed separately, anyone working on this?

This shouldn't be hard, `s/real/T/g` should do the trick. The question is how to transition the behaviour  properly.

**Comment by 9il** [comment]
> @9il I'm interested in learning more about this. Can't compilers inline a + b as well as a builtin would do?
> 
>  Revert View details

Plz, take a look
https://godbolt.org/g/xvEpNH

**Comment by 9il** [comment]
> dependance on phobos

The Phobos dependency is not a big problem. The problem that I can not modify phobos module for a target compiler like i can do with a dub package.

**Comment by 9il** [comment]
> It will need to work with -betterC (i.e. not depend on libdruntime.a)to have any hope of making Ilya happy. But assuming that all of the basic operations are pragma(inline, true), I don't think this will be a problem.

`std.math` that is used in `std.complex` is not betterC

**Comment by thewilsonator** [comment]
>Plz, take a look
https://godbolt.org/g/xvEpNH

I see only two calls to std.math.fabs in the entire dump. I was kinda expecting more.

**Comment by 9il** [comment]
> Plz, take a look
> https://godbolt.org/g/xvEpNH
> 
> I see only two calls to std.math.fabs in the entire dump. I was kinda expecting more.

Yes, I used only basic arithmetic ops. The problem is huge performance degradation.

**Comment by thewilsonator** [comment]
>The Phobos dependency is not a big problem. The problem that I can not modify phobos module for a target compiler like i can do with a dub package.

If this is to become `core.complex` as Andrei suggested  then I guarantee that LDC will allow you to optimise the LDC druntime implementation for as may arch's as you need to.

**Comment by wilzbach** [comment]
> Also druntime is not allowed to use phobos so its betterC issues will not apply.

Do we need to move it to `core`? We can also test that certain modules in Phobos are `-betterC`, see e.g.

https://github.com/dlang/phobos/pull/5952

**Comment by 9il** [comment]
> If this is to become core.complex as Andrei suggested then I guarantee that LDC will allow you to optimise the LDC druntime implementation for as may arch's as you need to.
> 
> Also druntime is not allowed to use phobos so its betterC issues will not apply.

Hmm, as you know `fastmath` is a set of some llvm flags. In practice we may need to mark function with one of other flags. But what flags should be choose `core./std.complex`? Should it be always `fastmath`? I should it be always not `@fastmath`? With native complex number we do not need to care about what kind of flags has library implementation. Native complex number inherit flags from the current function

**Comment by thewilsonator** [comment]
@andralex We should be able to have almost full backwards compatibility if we cannibalise the builtin names, the only thing we lose is the ability to have e.g. 0.0 + 0.0i as a valid initialiser (would become `cdouble(0.0, 0.0)`, I'm sure we can hack the lever/parser in the meantime.) and require people to `import core.complex;` everything else should be able to be kept the same.

**Comment by 9il** [comment]
> Do we need to move it to core? We can also test that certain modules in Phobos are -betterC, see e.g.
> 
> #5952

I do not think this is an issue. see https://github.com/dlang/phobos/pull/6014#issuecomment-358856178

**Comment by thewilsonator** [comment]
> In practice we may need to mark function with one of other flags. But what flags should be choose core./std.complex? Should it be always fastmath? I should it be always not @fastmath? With native complex number we do not need to care about what kind of flags has library implementation. Native complex number inherit flags from the current function

Perhaps we should add an `@inherit` attribute for inline functions?

**Comment by 9il** [comment]
> Perhaps we should add an @inherit attribute for inline functions?

Yes!!! But plus `__traits(llvmAttributes)` to chose algorithm depend on flags!
This would be awesome !

**Comment by 9il** [comment]
> Perhaps we should add an @inherit attribute for inline functions?
> 
> Yes!!! But plus __traits(llvmAttributes) to chose algorithm depend on flags!
> This would be awesome !

This will allow solve all problems I have with std.math and std.complex. And I will be happy to remove `mir.math.common` then

**Comment by thewilsonator** [comment]
>But plus `__traits(llvmAttributes)` to chose algorithm depend on flags!

Please elaborate, I'm not sure what this would do.

**Comment by 9il** [comment]
> But plus __traits(llvmAttributes) to chose algorithm depend on flags!
> 
> Please elaborate, I'm not sure what this would do.

`__traits(llvmAttributes)` returns an alias sequence of LLVM specific attributes that function has.

```d
@fastmath auto foo(double a); { return bar(a); }
@inherit auto bar(double a) {
   // this trivial case can be done by LLVM optimizer,
   // but helpful for other tricky math hacks
    static if (AllowsReciprocal!(__traits(llvmAttributes)))
        return PI * 1 / a; // one algo
    else
        return PI / a; // another algo
}

template AllowsReciprocal(Attrs...) { ... }
```

**Comment by thewilsonator** [comment]
That won't work because the body of `bar` must be independent of it's inlining into `foo` i.e. that static if has no knowledge of the inlining. You could template bar on a given set of attributes.

**Comment by 9il** [comment]
@thewilsonator can it be just a template without args? E. g. the attributes are special hiden template Park.

**Comment by wilzbach** [comment]
> We should be able to have almost full backwards compatibility

How would `core.complex` (i.e. `-betterC` `complex`) work?
`std.complex`
- has nice formatting (requires `std.format`)
- uses a lot of things from `std.math` which aren't `-betterC`

Wouldn't it then be just a copy of `std.complex` or in your words:

> I would need to copy std.complex to mir, mark with appropriate LDC flags

**Comment by thewilsonator** [comment]
Formating not so sure, any reason this can't be done one `std.format` as is done now?
Excluding the trig and `^^` and friends the only use of `std.math` is `fabs`.

**Comment by thewilsonator** [comment]
@9il I don't think so.

**Comment by ibuclaw** [comment]
> He means that std.complex is not annotated with LLVM optimisation hints (inlining, fast math). This leads to either duplication, or performance penalties and a dependance on phobos. Neither of which is desirable.

Neither are std.math or any other functions that use complex annotated with compiler specific attributes -> not a valid argument.

**Comment by ibuclaw** [comment]
> Perhaps we should add an @inherit attribute for inline functions? Edit: or make the ldc.attributes template inheritable like the builtin attributes.

If it's not part of the language, then I can't see it going in. Likewise fastmath, which means that code generated is no longer IEEE compliant.

**Comment by 9il** [comment]
> Neither are std.math or any other functions that use complex annotated with compiler specific attributes -> not a valid argument

There are no D libs for math and algorithms, but Phobos?

Entire mir-algorithm is annotated. I do not  whant to argue why it is fine for IEEE, you can do experiments with LDC to ensure that IEEE is preserved if a lambda is not annotated.

**Comment by 9il** [comment]
> If it's not part of the language, then I can't see it going in. Likewise fastmath, which means that code generated is no longer IEEE compliant.

It is part for the best D compiler for math and finance.

**Comment by stefan-koch-sociomantic** [comment]
@9il `AllowsReciprocal` ? 

when would that ever not be allowed ?
b/c of precision issues ?

**Comment by 9il** [comment]
@stefan-koch-sociomantic div operation can be optimised in fastmath to mul and reciprocal. This optimization is not allowed by default because of IEEE

**Comment by ibuclaw** [comment]
> Entire mir-algorithm is annotated. I do not whant to argue why it is fine for IEEE

You're free to do that if you value speed over correctness.

Talking about annotating functions with non-standard attributes is not up for debate though.

**Comment by ibuclaw** [comment]
> It is part of the best D compiler for math and finance.

The best compiler for math and finance is gdc (I say that with a grin, but the point stands - there is no best).

(incidentally one of the Berlin meet up talks did a live demo talk benchmarking math performance between dmd, ldc, gdc - the winner was gdc by a landslide, the loser was ldc but the author was surprised and didn't know why).

**Comment by stefan-koch-sociomantic** [comment]
gdc's frontend does a better job then ldc does.
also the gcc optimizers are capable of performing rewrites in more cases.
llvm's development is mostly driven by clang which emits rather specific patterns.
and in the optimizers only those patterns get recognized.

**Comment by 9il** [comment]
@ibuclaw this is your opinion. Mir-algorithm does not break IEEE, probably you do not understand how this attributes affects lambdas.

I do not debate. I just showed how the prs kill perfomance. I will be happy with my own library

**Comment by ibuclaw** [comment]
@9il I'll start by removing fastmath from your examples and seeing what the real difference is.

**Comment by thewilsonator** [comment]
>>Perhaps we should add an @inherit attribute for inline functions? Edit: or make the ldc.attributes template inheritable like the builtin attributes.
>If it's not part of the language, then I can't see it going in. Likewise fastmath, which means that code generated is no longer IEEE compliant.

@ibuclaw Quite the opposite.`@inherit` has nothing to do directly with IEEE conformance except propagating attributes of a function to functions inlined into it e.g.
 ```
@infer void bar(...) { ... } // <-not @fastmath
@fastmath void foo(...)
{
    bar(...); // inlined into foo. foo's inline copy of bar now @fastmath
}
```
 If there becomes a `core.complex` (and probably even if it stays as `std.complex`) LDC's implementation may add that attribute.

The question is if there is benefit in standardising it across compilers for propagating optimisation directives.

**Comment by ibuclaw** [comment]
> The question is if there is benefit in standardising it across compilers for propagating optimisation directives.

You could start with a DIP. ☺️

**Comment by thewilsonator** [comment]
Well would GDC benefit from this? DMD doesn't have much in the way of perf tweaking attributes.

**Comment by ibuclaw** [comment]
Try playing around with gdc yourself, and see if there might be: explore.dgnu.org

The module you're looking for is `gcc.attribute`, and the uda you should test is `@attribute("optimize", "-fsome-optimizer-flag")`. 

From what you describe though, I'm not sure.  I would expect all inlined functions to be applied the same optimization flags as the function they are inlined into without the need to be explicit about it.

**Comment by thewilsonator** [comment]
Well in LDC the IEEE compliance (or lack thereof) is encoded into the function such that inlining will result in some instructions being `@fastmath` and some not. Maybe it is done differently in GCC?

**Comment by ibuclaw** [comment]
In any case. Any solution or discussion needs to revolve around compiler-agnostic tests for performance and attributes. Anything else is out of scope.

We could introduce `core.complex`, but for what purpose? There is no complex in the language. So it will end up being a copy of std.complex.

**Comment by ibuclaw** [comment]
> Well in LDC the IEEE compliance (or lack thereof) is encoded into the function such that inlining will result in some instructions being @fastmath and some not. Maybe it is done differently in GCC?

I couldn't say off the top of my head. If anything it probably depends entirely on what order optimization passes are done in. I.e: whether or not inlining is done before or after reordering.

I'm pretty certain that optimisation attributes are applied only at a function level and not a block level. So I don't think ldc's problem really applies for gdc. But as I said please test this! (I'm on my phone and can't for the moment)

**Pull request merged**